packages/plugin-ext/src/common/rpc-protocol.ts

```diff
@@ -296,41 +291,45 @@ function canceled(): Error {
  *  - multiple messages to be sent from one stack get sent in bulk at `process.nextTick`.
  *  - each incoming message is handled in a separate `process.nextTick`.
  */
-class RPCMultiplexer implements Disposable {
+class RPCMultiplexer implements Disposable, MessageConnection {
 
     private readonly connection: MessageConnection;
     private readonly sendAccumulatedBound: () => void;
 
     private messagesToSend: string[];
 
+    private readonly messageEmitter = new Emitter<string>();
     private readonly toDispose = new DisposableCollection();
 
-    constructor(connection: MessageConnection, onMessage: (msg: string) => void, remoteHostId?: string) {
+    constructor(connection: MessageConnection) {
         this.connection = connection;
         this.sendAccumulatedBound = this.sendAccumulated.bind(this);
 
         this.toDispose.push(Disposable.create(() => this.messagesToSend = []));
-        this.toDispose.push(this.connection.onMessage((data: string[]) => {
+        this.toDispose.push(this.connection.onMessage((msg: string) => {
+            const data = JSON.parse(msg);
             const len = data.length;
             for (let i = 0; i < len; i++) {
-                onMessage(data[i]);
+                this.messageEmitter.fire(data[i]);
             }
```

Comment:
I feel like better naming could make it easier to understand the code here. Also, is there a reason to not use a `for ... of ...` construct?

```ts
const messages = JSON.parse(msg);
for (const message of messages) {
    this.messageEmitter.fire(message);
}
```

Reply:
>  is there a reason to not use a for ... of ... construct

No, other than that it was there before
> I feel like better naming could make it easier to understand the code here

Any concrete pointers what is badly named?

Reply:
What is `data`? In this case it's an array of messages because it accumulates messages to send in one shot. So like I proposed `messages` makes more sense. Then the `for ... of ...` makes it easy to read that we go through each. It just looks like obfuscated code for some reason with the way it is.